src/DataLayer/dataSources/DataSource.ts

```diff
@@ -1,25 +1,95 @@
+export const SERVER_SIDE_ROW_LIMIT = 1000;
+
+const ROW_DIMENSION_INDEX = 0;
+const COLUMN_DIMENSION_INDEX = 1;
+
 export type ExecFactory<T> = (resultSpec: AFM.IResultSpec) => Promise<T>;
 
 export class DataSource<T> implements IDataSource<T> {
+    private executionPromises: {
+        [key: string]: Promise<Execution.IExecutionResponse>
+    };
+
     constructor(
         private execFactory: ExecFactory<T>,
         private afm: AFM.IAfm,
-        private fingerprint?: string
-    ) {}
+        private fingerprint?: string,
+        private responseFactory?: (resultSpec: AFM.IResultSpec) => Promise<Execution.IExecutionResponse>,
+        private resultFactory?:
+            (executionResultUri: string, offset: number[], limit: number[]) => Promise<Execution.IExecutionResult>
+    ) {
+        this.executionPromises = {};
+    }
 
     public getData(resultSpec: AFM.IResultSpec): Promise<T> {
         return this.execFactory(resultSpec);
     }
 
+    public getPage(
+        resultSpec: AFM.IResultSpec,
+        offset: number[] = [],
+        limit: number[] = []
+    ): Promise<Execution.IExecutionResponses> {
+        const resultSpecFingerprint = stringify(resultSpec);
+        if (!this.responseFactory) {
+            throw new Error('Missing responseFactory!');
+        }
+        if (!this.executionPromises[resultSpecFingerprint]) {
+            this.executionPromises[resultSpecFingerprint] = this.responseFactory(resultSpec);
+        }
+
+        return this.executionPromises[resultSpecFingerprint].then(
+            (executionResponse: Execution.IExecutionResponse) => {
+                if (!this.resultFactory) {
+                    throw new Error('Missing resultFactory!');
+                }
+                const [ offsetRows = 0, offsetColumns = 0 ] = offset;
+
+                const rowLimit = limit[ROW_DIMENSION_INDEX];
+                const safeRowLimit = typeof rowLimit === 'number'
+                    && (rowLimit < SERVER_SIDE_ROW_LIMIT)
+                    ? rowLimit
+                    : SERVER_SIDE_ROW_LIMIT;
+
+                const columnLimit = limit[COLUMN_DIMENSION_INDEX];
+                const safeColumnLimit = typeof columnLimit === 'number'
+                    && (columnLimit < SERVER_SIDE_COLUMN_LIMIT)
+                    ? columnLimit
+                    : SERVER_SIDE_COLUMN_LIMIT;
+
+                if (rowLimit > safeRowLimit) {
+                    // tslint:disable-next-line:no-console
+                    console.warn('The maximum row limit per page is ' + SERVER_SIDE_ROW_LIMIT);
+                }
+                if (columnLimit > safeColumnLimit) {
+                    // tslint:disable-next-line:no-console
+                    console.warn('The maximum column limit per page is ' + SERVER_SIDE_COLUMN_LIMIT);
+                }
+
+                return this.resultFactory(
+                    executionResponse.links.executionResult,
+                    [offsetRows, offsetColumns],
```

Comment:
are you going to always request with two dimensional data? 

this looks like a dangerous idea